### PR TITLE
[TRTLLM-12671][chore] Refactor visual_gen's attention handling; Fix LTX-2 & Cosmos3

### PR DESCRIPTION
**Pull request by xrq-phys**
## Dev Engineer Review

- Refactors VisualGen attention to use explicit cross-attention flags and per-site `AttentionMetadata`.
- Moves metadata construction to transformers and metadata reuse and CUDA graph registration to pipelines.
- Fixes SageAttention cross-attention KV-layout and KV-cache handling.
- Updates FLUX, Cosmos3, LTX-2, Qwen Image, and Wan APIs and execution paths.
- Removes obsolete `attention_metadata_state` handling.
- Review focus: API consistency, metadata validation, cross-attention lengths, CUDA graph keys, and sequence-parallel behavior.
- No configuration or test-list changes were detected.

## QA Engineer Review

- Updates unit and multi-GPU tests for attention metadata propagation across VisualGen backends and pipelines.
- Adds coverage for metadata construction, validation, buffer reuse, CUDA graph constraints, cross-attention, SageAttention, and sequence parallelism.
- Adds tests including:
  - `test_self_attention_site_is_not_cross`
  - `test_mixed_site_carries_distinct_kv_length`
  - `test_no_kv_cache_path`
  - `test_reprepare_keeps_device_buffers_pointer_stable`
  - `test_batch_size_change_is_supported`
  - `test_batch_size_beyond_capacity_raises`
  - `test_seq_lens_batch_mismatch_raises`
  - `test_base_metadata_type_works_for_backends_that_ignore_metadata`
- Removes `test_trtllm_attention_metadata.py`.
- No `tests/integration/test_lists/`, `test-db/`, `qa/`, or `waives.txt` changes were detected.
- Coverage mapping for the changed test functions is unavailable.
- Verdict: needs follow-up.

## Description

- Fix SageAttention cross attention
  * Separate is_cross definition in `thop` vs. in python: `thop` needs specialized is_cross treatment only if KV cache is paged or ragged, while `separateQkv` cross routes through regular context attention.
  * Fix SageAttention invalid memory at cross-attn: derive KV size from the right params (`kv_len_...` was delicated to generation. use `total_kv_...` instead)
  * Depends on #17648 to fully fix for any KV sequence.
- Metadata handling: align with LLM
  * Assign separate attention metadata to each attention site e.g. in Wan-like models, self-attention and cross-attention will use different metadata objects.
  * Each transformer class implementation will expose a `create_attn_metadata` to create metadata objects for each of its attention sites
  * Pipeline will create, maintain, and reuse attention metadata objects throughout the denoising process.
  * Pipeline will register metadata as CUDA graph keys.
- Cleanup VisualGen attention modules
  * Resolve cross-attention definition: never infer cross/self attention from `qkv_format`: Use an explicit is_cross flag.
  * `qkv_format` only indicates whether the module uses fused `to_qkv` or separate `to_q, to_k, to_v`

## Test Coverage

- `tests/unittest/_torch/visual_gen/test_attention_metadata.py` (new): site construction, `is_cross` identity semantics, pointer-stable re-prepare, batch changes, and that construction is legal outside but not inside a graph capture.
- Existing visual_gen unit and pipeline suites updated for the new signatures.
- SageAttention cross-attention verified against VANILLA across a q x kv sweep at 8 and 16 heads: previously every `kv > q` case faulted or returned cos ~0.80, now all are ~0.999, including a 16x ratio and the minimal one-block overrun that used to corrupt silently.
- Cosmos3-Nano t2v end-to-end with SageAttention.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by xrq-phys** [comment]
@coderabbitai summary

File: tests/unittest/_torch/visual_gen/test_flux_pipeline.py [hunk 1]
```diff
@@ -95,6 +96,7 @@ def _get_flux_transformer_inputs(transformer, device="cuda", dtype=torch.bfloat1
 
     return dict(
         hidden_states=hidden_states,
+        attn_metadata=flux_attn_metadata(transformer, hidden_states, encoder_hidden_states),
```
[the remark this answers is not in the document]
Comment (xrq-phys) [reply tests/unittest/_torch/visual_gen/test_flux_pipeline.py:99]: Regenerate `attn_metadata` breaks dynamo tracing & CUDA graph. Did you take that into account? Did you reference LLM's metadata handling when reviewing this?

[the remark this answers is not in the document]
Comment (xrq-phys) [reply tests/unittest/_torch/visual_gen/test_flux_pipeline.py:99]: metadata is only used by the attention backend wrapped **under** ulysses. Why do you insist I need a rank-local metadata?

**Comment by xrq-phys** [comment]
@coderabbitai  Cleanup broken comments on "review status"

**Comment by xrq-phys** [comment]
@coderabbitai summary

**Comment by xrq-phys** [comment]
Replaced PR description with a handwritten one.

File: tensorrt_llm/_torch/visual_gen/modules/attention.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ class QKVMode(str, Enum):
     SEPARATE_QKV = "separate"
 
 
-# TODO: torch compile
+# FIXME: does trtllm offer fused routine for this via torch.ops.trtllm?
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/modules/attention.py:24]: I dont think it is currently available. Normally we would fuse qk rope w/ its producer such as qk_norm

File: tensorrt_llm/_torch/visual_gen/attention_backend/metadata.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    else:
+        tensor = torch.tensor(list(value), dtype=torch.int32)
+
+    if tensor.shape[0] != batch_size:
+        raise ValueError(f"sequence lengths have batch {tensor.shape[0]} but {batch_size=}")
+    return tensor
+
+
+@torch.compiler.disable
+def _cu_seqlens(seq_lens: torch.Tensor, attn_metadata: AttentionMetadata) -> torch.Tensor:
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/attention_backend/metadata.py:58]: pls add some doc-string

File: tensorrt_llm/_torch/visual_gen/modules/attention.py [hunk 4]
```diff
@@ -515,6 +538,7 @@ def _attn_impl(
         q: torch.Tensor,
         k: torch.Tensor,
         v: torch.Tensor,
+        attn_metadata: AttentionMetadata,
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/modules/attention.py:541]: Blocking: this makes attn_metadata mandatory, but HunyuanVideo 1.5 was not migrated. Its two Attention subclasses still call _attn_impl(q, k, v, ...) without metadata (transformer_hunyuan_video1_5.py:318 and :649), and its pipeline never creates or passes metadata. That registered pipeline now raises TypeError before the backend runs. Please migrate Hunyuan as part of this contract change and add a registry-wide audit/test so another supported model cannot be missed.

File: tensorrt_llm/_torch/visual_gen/models/modeling.py [hunk 5]
```diff
@@ -36,12 +78,34 @@ def __init__(self, model_config: DiffusionModelConfig):
...
+    def attn_backend_metadata_cls(self) -> type[AttentionMetadata]:
+        """Metadata type this component's attention backend expects."""
+        return get_visual_gen_attention_backend(self.model_config.attention.backend).Metadata
+
+    @property
+    def attn_requires_metadata(self) -> bool:
+        """Whether any attention site in this model needs prepared metadata."""
+        from tensorrt_llm._torch.visual_gen.modules.attention import Attention
+
+        return any(m.requires_metadata for m in self.modules() if isinstance(m, Attention))
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/models/modeling.py:91]: Can we derive the metadata capability and type from each effective Attention site instead of combining a model-wide any() with the configured backend class? Routing is per instance: cross attention can fall back to VANILLA while self attention stays TRTLLM. The current API therefore loses the per-site backend contract and makes every model repeat capability checks. A shared builder could take site -> (Attention instance, Q/KV lengths), inspect requires_metadata and the selected backend metadata type, and let the model provide only the shape semantics.

File: tensorrt_llm/_torch/visual_gen/models/wan/transformer_wan.py [hunk 6]
```diff
@@ -801,9 +821,81 @@ def unpatchify(self, x, original_shape):
...
+
+        q_self, kv_self = get_ulysses_seq_lens(s_video, s_video, visual_gen_mapping=vgm)
+        sites = {
+            "self": make_diffusion_attn_metadata(
+                metadata_cls,
+                batch_size=batch_size,
+                q_seq_lens=q_self,
+                kv_seq_lens=None if kv_self == q_self else kv_self,
+            ),
+            "cross_text": make_diffusion_attn_metadata(
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/models/wan/transformer_wan.py:878]: Non-blocking (fine as a follow-up): could the site keys be typed rather than free-form strings? e.g. a shared `AttnSite(q=Stream.VIDEO, kv=Stream.TEXT)` descriptor with `is_cross` derived from `q != kv`, so `self`/`cross_text`/`video_to_audio`/`mixed` all become the same vocabulary and get validated once in the base `create_attn_metadata()`?
